# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/indexedDB/ActorsParent.cpp
# Commit: 09e4368e2355
# Full Hash: 09e4368e23552498b4521f17452a73223b935d3e
# Author: Simon Giesecke <sgiesecke@mozilla.com>
# Date: 2019-12-05 04:11:08
# Regressor Bug: 1600906
# File Overlap Count: 1
# Description:
#   Bug 1600906 - Make use of FlippedOnce in FullObjectStoreMetadata. r=dom-workers-and-storage-reviewers,ytausky
#   
#   Depends on D55676
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D55677
# ==============================================================================

diff -r fcea7ad7ad7b -r 09e4368e2355 dom/indexedDB/ActorsParent.cpp
--- a/dom/indexedDB/ActorsParent.cpp	Wed Dec 04 14:30:55 2019 +0000
+++ b/dom/indexedDB/ActorsParent.cpp	Wed Dec 04 14:26:48 2019 +0000
@@ -363,26 +363,18 @@
 
 typedef nsRefPtrHashtable<nsUint64HashKey, FullIndexMetadata> IndexTable;
 
+// Can be instantiated either on the QuotaManager IO thread or on a
+// versionchange transaction thread. These threads can never race so this is
+// totally safe.
 struct FullObjectStoreMetadata {
-  ObjectStoreMetadata mCommonMetadata;
+  ObjectStoreMetadata mCommonMetadata = {0, nsString(), KeyPath(0), false};
   IndexTable mIndexes;
 
   // These two members are only ever touched on a transaction thread!
-  int64_t mNextAutoIncrementId;
-  int64_t mCommittedAutoIncrementId;
-
-  bool mDeleted;
-
- public:
-  FullObjectStoreMetadata()
-      : mCommonMetadata(0, nsString(), KeyPath(0), false),
-        mNextAutoIncrementId(0),
-        mCommittedAutoIncrementId(0),
-        mDeleted(false) {
-    // This can happen either on the QuotaManager IO thread or on a
-    // versionchange transaction thread. These threads can never race so this is
-    // totally safe.
-  }
+  int64_t mNextAutoIncrementId = 0;
+  int64_t mCommittedAutoIncrementId = 0;
+
+  FlippedOnce<false> mDeleted;
 
   NS_INLINE_DECL_THREADSAFE_REFCOUNTING(FullObjectStoreMetadata);
 
@@ -14917,12 +14909,12 @@
     return IPC_FAIL_NO_REASON(this);
   }
 
-  foundMetadata->mDeleted = true;
+  foundMetadata->mDeleted.Flip();
 
   DebugOnly<bool> foundTargetId = false;
   const bool isLastObjectStore = std::all_of(
       dbMetadata->mObjectStores.begin(), dbMetadata->mObjectStores.end(),
-      [&foundTargetId, aObjectStoreId](const auto& objectStoreEntry) {
+      [&foundTargetId, aObjectStoreId](const auto& objectStoreEntry) -> bool {
         if (uint64_t(aObjectStoreId) == objectStoreEntry.GetKey()) {
           foundTargetId = true;
           return true;
